Fix font-variant shorthand serialization

nt1m opened the pull request:
#### cee0d60d4af3a42f1dfecc60aeb3928df9d7d2c1
<pre>
Fix font-variant shorthand serialization
<a href="https://bugs.webkit.org/show_bug.cgi?id=247919">https://bugs.webkit.org/show_bug.cgi?id=247919</a>
rdar://102339716

Reviewed by Darin Adler.

- Fixed bug where font-variant-ligatures: none + another non-default longhand would serialize even if it not possible
- Clean up fontValue()/fontVariantValue() to not consider implicit/explicit anymore used to represent default/system values
- Refactor to remove appendFontLonghandValueIfExplicit()

* LayoutTests/fast/text/font-variant-shorthand-expected.txt:
* LayoutTests/fast/text/font-variant-shorthand.html:
* LayoutTests/fast/css/font-property-priority-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/font-shorthand-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/font-variant-shorthand-serialization-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/font-variant-shorthand-serialization.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/font-shorthand-serialization-expected.txt:
* Source/WebCore/css/CSSProperties.json:
Updated ordering to match order from the spec (which matters for serialization).

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::fontValue const):
Inlined parts of appendFontLonghandValueIfExplicit, simplified font-stretch handling.

(WebCore::StyleProperties::fontVariantValue const):
Rewrite to stop using appendFontLonghandValueIfExplicit, fix various serialization issues when collapsing longhands.

(WebCore::canUseShorthandForLonghand):
Add notes on why collapsing is not possible.

(WebCore::StyleProperties::appendFontLonghandValueIfExplicit const): Deleted.

* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeFontVariantShorthand):

Canonical link: <a href="https://commits.webkit.org/256681@main">https://commits.webkit.org/256681@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ac000242265a4977b7233ba0a74988829b085750

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96513 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5767 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29586 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106044 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/166398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100495 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5951 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34510 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88886 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/102768 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102186 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83116 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31406 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/86284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/88171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40241 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19662 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37909 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/21056 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4634 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4176 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/991 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40327 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -685,16 +650,44 @@ String StyleProperties::textDecorationSkipValue() const
+            return emptyString(); // All longhands must have at least implicit values.
+
+        auto value = propertyAt(foundPropertyIndex).value();
+        auto keyword = valueID(value);
```
Should use `propertyAsValueID`.

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -685,16 +650,44 @@ String StyleProperties::textDecorationSkipValue() const
+
+        // If all properties are set to the same special keyword, serialize as that.
+        // If some but not all properties are, the font shorthand can't represent that, serialize as empty string.
+        if (isCSSWideKeyword(keyword) || CSSPropertyParserHelpers::isSystemFontShorthand(keyword)) {
```
This `isSystemFontShorthand` check is not quite right. Unlike the `fontValue` function, the `fontVariantValue` function should **never** return a system font name, instead should serialize as an empty string if any of the longhands have the system font name. We should add a test that checks this if we don’t already have one.

nt1m replied:
Changed to return an empty string. Though we have the same problem with serializing font-family/size, as one of the WPT shows.

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -551,36 +512,40 @@ String StyleProperties::fontValue() const
+            return; // All longhands must have at least implicit values if "font" is specified.
+
+        if (propertyAt(foundPropertyIndex).isImplicit())
+            return;
```
Later we need to change this to be based on the value of the properties, not on `isImplicit`.

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -551,36 +512,40 @@ String StyleProperties::fontValue() const
 
-    // This code no longer uses commonValue, for now we define it so we can use appendFontLonghandValueIfExplicit.
-    String commonValue;
+    auto appendValueIfExplicit = [this](CSSPropertyID property, StringBuilder& result) {
```
Why not define this after defining the `StringBuilder` and just capture it?

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -551,36 +512,40 @@ String StyleProperties::fontValue() const
+        const char* prefix = property == CSSPropertyLineHeight ? " / " : " ";
+
+        String value = propertyAt(foundPropertyIndex).value()->cssText();
+        result.append(result.isEmpty() ? "" : prefix, value);
```
I would not bother with the local variable, just write this all on one line.

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -551,36 +512,40 @@ String StyleProperties::fontValue() const
+        if (propertyAt(foundPropertyIndex).isImplicit())
+            return;
+
+        const char* prefix = property == CSSPropertyLineHeight ? " / " : " ";
```
I personally would use `auto` instead of `const char*`.

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -551,36 +512,40 @@ String StyleProperties::fontValue() const
+
+        const char* prefix = property == CSSPropertyLineHeight ? " / " : " ";
+
+        String value = propertyAt(foundPropertyIndex).value()->cssText();
```
We do not need to call `propertyAt` twice, let’s use a local variable for it.

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -551,36 +512,40 @@ String StyleProperties::fontValue() const
+    appendValueIfExplicit(CSSPropertyFontStyle, result);
+    appendValueIfExplicit(CSSPropertyFontVariantCaps, result);
+    appendValueIfExplicit(CSSPropertyFontWeight, result);
+    if (stretchKeyword && stretchKeyword != CSSValueNormal)
```
I suggest:
```
if (stretchKeyword && *stretchKeyword != CSSValueNormal)
```
or
```
if (stretchKeyword.value_or(CSSValueNormal) != CSSValueNormal)
```

darinadler reviewed (approved):
This fixes bugs, but there are no tests showing the bugs we are fixing. Can we add some WPT test coverage for all the problems this fixes?

nt1m committed:
Fix font-variant shorthand serialization
https://bugs.webkit.org/show_bug.cgi?id=247919
rdar://102339716

Reviewed by Darin Adler.

- Fixed bug where font-variant-ligatures: none + another non-default longhand would serialize even if it not possible
- Clean up fontValue()/fontVariantValue() to not consider implicit/explicit anymore used to represent default/system values
- Refactor to remove appendFontLonghandValueIfExplicit()

* LayoutTests/fast/text/font-variant-shorthand-expected.txt:
* LayoutTests/fast/text/font-variant-shorthand.html:
* LayoutTests/fast/css/font-property-priority-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/font-shorthand-serialization-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/font-variant-shorthand-serialization-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/font-variant-shorthand-serialization.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/font-shorthand-serialization-expected.txt:
* Source/WebCore/css/CSSProperties.json:
Updated ordering to match order from the spec (which matters for serialization).

* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::fontValue const):
Inlined parts of appendFontLonghandValueIfExplicit, simplified font-stretch handling.

(WebCore::StyleProperties::fontVariantValue const):
Rewrite to stop using appendFontLonghandValueIfExplicit, fix various serialization issues when collapsing longhands.

(WebCore::canUseShorthandForLonghand):
Add notes on why collapsing is not possible.

(WebCore::StyleProperties::appendFontLonghandValueIfExplicit const): Deleted.

* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::consumeFontVariantShorthand):

Canonical link: https://commits.webkit.org/256681@main
